chore: add local-npm tool

izaera opened the pull request:
Please see the README in the commit to know what all this is about...

bryceosterhaus commented on support/packages/local-npm/README.md:
```diff
@@ -0,0 +1,213 @@
+3. `registry`: this has two subcommands for querying (`get`) the active
+   repository, or setting (`set`) it to `local` or `public`.
+
+Usually you will always have your `npm` and `yarn` repos pointing to your local
+repo. This is because Verdaccio acts as a proxy of the real public repo so, if
+you request anything it doesn't have, it will download it, cache it and serve
+it.
```
```suggestion
Usually you will always have `npm`/`yarn` pointing to your local
registry. This is because Verdaccio acts as a proxy of the public registry so, if
you request anything it doesn't have, it will download it, cache it and serve
it.
```

bryceosterhaus commented on support/packages/local-npm/README.md:
```diff
@@ -0,0 +1,213 @@
+you request anything it doesn't have, it will download it, cache it and serve
+it.
+
+This has the benefit that you don't need to switch between local and public
+repos whenever you use third party packages and, also, that you will have a
+local mirror of the public `npm` repo which will let you develop even when
+offline (this is not a wanted feature of the setup, but a nice side effect).
```
```suggestion
This has the benefit that you don't need to switch between local and public
registries whenever you use third party packages and you will have a
local mirror of the public `npm` registry which will let you develop even when
offline (this is not a wanted feature of the setup, but a nice side effect).
```

bryceosterhaus commented on support/packages/local-npm/README.md:
```diff
@@ -0,0 +1,213 @@
+local mirror of the public `npm` repo which will let you develop even when
+offline (this is not a wanted feature of the setup, but a nice side effect).
+
+The only moments when you need to switch to using the public repos are:
+
+1. If you are going to update a `yarn.lock` file that you will then commit to a
+   GitHub repo.
+2. If you want to make sure that a public release has gone well and thus you
+   need the possibility that your local repo serves packages that are not really
+   available to the public.
```
```suggestion
The only moments when you need to switch back to using the public registry are:

1. If you are going to update a `yarn.lock` file that you will commit to upstream.
2. If you want to make sure that a public release has worked.
```

bryceosterhaus commented on support/packages/local-npm/README.md:
```diff
@@ -0,0 +1,213 @@
+
+#### Case 1
+
+Case 1 will usually happen at the end of a pull request preparation. Whenever we
+change deps in `liferay-frontend-projects`, we must run `yarn install` so that
+`yarn.lock` is updated and we can commit the modified version.
+
+However, `yarn.lock` files hold information about the repo from where the
+packages are downloaded so, if you run `yarn install` when your machine is
+pointing to your local repo, the `yarn.lock` files will have references to
+`localhost` instead of `registry.npmjs.org`, and those will fail in other
+machines.
+
+For that reason, the recommended action to take when you finish preparing your
+pull request are:
+
+1. Run `git checkout yarn.lock` to restore it to `master`'s state
+2. Run `local-npm registry set public` to point `npm` and `yarn` to the public
+   repo.
+3. Run `yarn install` to update `yarn.lock`
+4. Run `local-npm registry set local` to restore your repos back again to use
+   your local Verdaccio.
+
+In the event that you forget to follow these steps or commit a `yarn.lock` file
+with references to `localhost` there's no need for drama (though you can make
+drama of it if you like; it's OK) because our CI build will catch the error and
+tell you that your commited `yarn.lock` is incorrect.
+
+In that case, simply fix it, force push the `yarn.lock` commit, and continue
+with you beautiful life.
```
```suggestion
Case 1 will usually happen at the end of a pull request preparation. Whenever we
change dependencies in `liferay-frontend-projects`, we must run `yarn install` so that
`yarn.lock` is updated and we can commit the modified version.

However, `yarn.lock` files hold information about the registry from where the
packages are downloaded. If you run `yarn install` when your machine is
pointing to your local registry, the `yarn.lock` files will have references to
`localhost` instead of `registry.npmjs.org`, and those will fail in other
machines.

For that reason, the recommended action to take when you finish preparing your
pull request are:

1. Run `git checkout yarn.lock` to restore it to `master`'s state
2. Run `local-npm registry set public` to point `npm` and `yarn` to the public
   repo.
3. Run `yarn install` to update `yarn.lock`
4. Run `local-npm registry set local` to restore your repos back again to use
   your local registry.

In the event that you forget to follow these steps or commit a `yarn.lock` file
with references to `localhost` there's no need for drama (though you can make
drama of it if you like; it's OK) because our CI build will catch the error and
tell you that your commited `yarn.lock` is incorrect.

In that case, simply fix it, force push the `yarn.lock` commit, and continue
with you beautiful life.
```

bryceosterhaus commented on support/packages/local-npm/README.md:
```diff
@@ -0,0 +1,213 @@
+
+Case 2 would arise after merging the PR and releasing your work to the public.
+In that moment, if you want to make sure that everything is correct and people
+outside won't have any problem, you may point to the `public` repo and do 
```
```suggestion
outside won't have any problem, you may point to the `public` registry and do 
```

bryceosterhaus commented on support/packages/local-npm/README.md:
```diff
@@ -0,0 +1,213 @@
+
+Perico has two options:
+
+1. Invoke `yarn link` on every test project he generates to see how
```
```suggestion
1. Use `yarn link` on every test project he generates to see how
```

bryceosterhaus commented on support/packages/local-npm/README.md:
```diff
@@ -0,0 +1,213 @@
+diagnostic code and see it work in the test project.
+
+Once everything is fixed, Perico prepares the PR, and sends it to GH.
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
```
We can delete this?

Maybe add an ending section about feedback, issues, etc

bryceosterhaus commented on support/packages/local-npm/bin/local-npm.js:
```diff
@@ -0,0 +1,47 @@
+	)
+	.command(
+		['publish [<projects>...]', 'p'],
+		'Publish current project or a list of projects to the local repo'
```
```suggestion
		'Publish current project or a list of projects to the local registry'
```

bryceosterhaus commented on support/packages/local-npm/bin/local-npm.js:
```diff
@@ -0,0 +1,47 @@
+	)
+	.command(
+		['registry', 'r'],
+		'Set/get the current npm and yarn configured repositories',
```
```suggestion
		'Set/get the current npm and yarn configured registries',
```

bryceosterhaus commented on support/packages/local-npm/bin/local-npm.js:
```diff
@@ -0,0 +1,47 @@
+		'Set/get the current npm and yarn configured repositories',
+		(yargs) =>
+			yargs
+				.command(['get', 'g'], 'Print current npm/yarn repo')
```
```suggestion
				.command(['get', 'g'], 'Print current npm/yarn registry')
```

bryceosterhaus commented on support/packages/local-npm/lib/index.js:
```diff
@@ -0,0 +1,190 @@
+const fs = require('fs');
+const path = require('path');
+
+const LOCAL_REGISTRY = 'http://localhost:4873';
```
Maybe add a comment on why this port is choesen. Pointing back to verdaccio's docs would probably be enough

izaera replied:
This really depends on how we install Verdaccio, so it should probably be configurable. But I wanted to see first how people decides/tries to install Verdaccio, then decide what to do with this....

bryceosterhaus commented on support/packages/local-npm/lib/index.js:
```diff
@@ -0,0 +1,190 @@
+
+	if (npmUrl === NPM_REGISTRY || yarnUrl === YARN_REGISTRY) {
+		console.log('');
+		console.log('  🔴 NOT using LOCAL repo: refusing to do anything');
```
```suggestion
		console.log('  🔴 NOT using LOCAL registry: refusing to do anything');
```

bryceosterhaus commented on support/packages/local-npm/lib/index.js:
```diff
@@ -0,0 +1,190 @@
+
+	projects = projects.map((name) => path.resolve(name));
+
+	console.log('');
```
these empty logs in this function seem odd. Can we log a newline or something more explicit? Otherwise I suspect they might get SF removed

bryceosterhaus commented on support/packages/local-npm/lib/index.js:
```diff
@@ -0,0 +1,190 @@
+
+	console.log('');
+
+	if (npmUrl === NPM_REGISTRY && yarnUrl === YARN_REGISTRY) {
+		console.log('  🌍 Using PUBLIC repo');
+	}
+	else if (npmUrl === NPM_REGISTRY) {
+		console.log("  🔴 Using public npm repo, but local yarn (that's bad)");
+	}
+	else if (yarnUrl === YARN_REGISTRY) {
+		console.log("  🔴 Using local npm repo, but remote yarn (that's bad)");
+	}
+	else {
+		console.log('  🏠 Using LOCAL repo');
+	}
+
+	console.log('');
```
```suggestion
	if (npmUrl === NPM_REGISTRY && yarnUrl === YARN_REGISTRY) {
		console.log('  🌍 Using PUBLIC registry');
	}
	else if (npmUrl === NPM_REGISTRY) {
		console.log("  🔴 Using public npm registry, but local yarn (that's bad)");
	}
	else if (yarnUrl === YARN_REGISTRY) {
		console.log("  🔴 Using local npm registry, but remote yarn (that's bad)");
	}
	else {
		console.log('  🏠 Using LOCAL registry');
	}

	console.log('');
```

bryceosterhaus reviewed (changes requested):
Great starting place for these docs and I like the example workflow you included.

I added a handful of suggestions, mostly just wordsmithing a bit. One thing I would re-check in the entire PR is the use of "repo" vs "registry". I think I caught most instances, but I may have missed a few...

kresimir-coko reviewed (approved):
This looks game changing @izaera great job!

izaera commented on support/packages/local-npm/README.md:
```diff
@@ -0,0 +1,213 @@
+# @liferay/local-npm
+
+`@liferay/local-npm` is a private package internal to this monorepo that
+provides a CLI tool to leverage local npm repos (like 
+[verdaccio](https://verdaccio.org/)) to publish local versions of development
+packages to be able to test them before the public release is made.
+
```
:thinking: It doesn't setup verdaccio, but uses it...

Maybe:

> `@liferay/local-npm` [...] provides a CLI tool to be able to use local npm repos (like 
> [verdaccio](https://verdaccio.org/)) to publish local versions of development
> packages so that they can be tested before the public release is made.

:thinking:

bryceosterhaus replied:
Yeah thats a good change. I hadn't gotten to the "how to use" part yet, so I was sort of guessing that it started it for you lol

izaera commented on support/packages/local-npm/README.md:
```diff
@@ -0,0 +1,213 @@
+Simply run `yarn link` inside the project's folder to make the `local-npm` CLI
+executable available in your system.
+
+> 👀 Note that we don't publish versions of this package to the public npm
+> registry, so the only way to install it is to `yarn link` it.
```
Nope, you can link anything you want.

The only problem that can arise if you use a non-published package as a dependency of a project is that every `yarn install` you run will fail (even if you have linked it prior to that install). That, obviously, blocks you unless you remove the offending package from `package.json` to be able to install the other dependencies first, then link what you need.

But in this case, we are not affected by that because we don't put this package as a dependency of anything. We just run `yarn link` to make the tool available in the OS's PATH, so that we can run it from the CLI. Think of it as a local `npm install -g`...

izaera commented:
Hey @bryceosterhaus, thx a bunch for the corrections :raised_hands: 

I've merged all proposed changes, except for [this](https://github.com/liferay/liferay-frontend-projects/pull/706#discussion_r732947133), which I changed a bit and added as the last commit.

I also searched for instances of repo vs registries and changed some more.

I'll merge this so that people can start using it. Feel free to ping me if you see anything else that needs to be changed, and I'll send a followup PR :+1:
